### PR TITLE
Add organization or facility-specific patient registration link capability to the backend

### LINKED ISSUE #1349: Support org-specific (+ facility specific) links for patient registration
Example: https://www.simplereport.gov/register/orgname/facilityName

^ It will be _easier_ to do https://www.simplereport.gov/app/register/orgname/facilityName but a redirect w/o `app` makes for a good nice-to-have

- [ ] Create organization (+ optional facility) specific links for registering patients
- [ ] Backend- have concept of human readable link

Figma: https://www.figma.com/file/CEzBM8kUw8LyWdfAhUThYb/Engineering-Ready-Prototype?node-id=2202%3A29831

### PR DESCRIPTION
**Pull request by nickscialli-usds**
## Related Issue or Background Info

- Resolves #1349 

## Changes Proposed

- Adds a migration for a new `patient_registration_link` table
- Adds `PatientRegistrationLink` model and associated repository
- Adds some tests

## Additional Information

- For the purpose of easily ensuring uniqueness, this seemed to make sense as its own table rather than an additional field on both the org and facility tables
- Added a constraint to this new table using an `sql` field in the migration file to make sure that each row has either an org ID or facility ID but not both or neither. (I didn't test this constraint; is there a good way to test?).

### UI
- [NA] Any changes to the UI/UX have been approved by design
- [NA] Any new or edited error messages have been approved by design 

### Testing
- [X] Includes a summary of what a code reviewer should verify

### Changes are Backwards Compatible
- [X] Database changes are submitted as a separate PR (this is the PR)
- [NA] GraphQL schema changes are backward compatible with older version of the front-end

### Security
- [NA] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [NA] Any dependencies introduced have been vetted and discussed

## Cloud
- [NA] DevOps team has been notified if PR requires ops support

### CODE REVIEW

File: backend/src/main/resources/db/changelog/db.changelog-master.yaml [hunk 1]
```diff
@@ -1758,3 +1758,52 @@ databaseChangeLog:
...
+                    unique: true
+                    foreignKeyName: fk__link__facility
+                    references: facility
+              - column:
+                  name: organization_id
+                  type: *idtype
+                  remarks: The organization associated with the link
+                  constraints:
+                    nullable: true
+                    unique: true
```
Comment (benwarfield-usds) [inline backend/src/main/resources/db/changelog/db.changelog-master.yaml:1797]: Definitely not true.

Comment (nickscialli-usds) [reply backend/src/main/resources/db/changelog/db.changelog-master.yaml:1797]: I believe the way I have designed this you'd never have an organization ID in this table more than once. If the `patient_registration_link` maps to a facility, that row would only have a `facility_id` and if the `patient_registration_link` maps to an organization, that row would only have an `organization_id`. So I _think_ the `unique` constraint holds?

Comment (nullflux) [reply backend/src/main/resources/db/changelog/db.changelog-master.yaml:1797]: I don't see what's broken about this either...?

**Review by nullflux — APPROVED** [review]
I originally expected that this PR would include an endpoint in the PXP Controller to retrieve the organization or facility name based on registration link, but that wasn't at all clear in the originating ticket and can wait until where that behavior is specified in #1412 

That said, LGTM!

File: backend/src/main/resources/db/changelog/db.changelog-master.yaml (hunk 1 above)
Comment (benwarfield-usds) [reply backend/src/main/resources/db/changelog/db.changelog-master.yaml:1797]: Sorry, I lost track of this conversation.

It seems like I'm confused about the expected usage pattern here—my initial assumption was that these were expirable links, but it seems like they're a durable optional property of the organization or facility, which means I'm wrong about the uniqueness. (There are exciting and unpleasant things to discover about unique indexes and nullable fields, but I think this pattern mostly avoids them, with the exception of some possible minor performance issues that don't matter for this scale.)

I'm still not _not_ confused about the purpose of the entity, but I agree that I was mistaken about the constraint being wrong, and should have responded more swiftly to the correction. Sorry!

**Comment by nickscialli-usds** [comment]
@nullflux I initially misread the ticket and now I'm wondering if this is the right approach. The ticket specifies that you might have an organization registration link that looks like `register/org-slug` or you might have a facility registration link that looks liek `register/org-slug/facility-slug`.

The way I coded this ticket was that it would look more like `register/org-or-facility-slug`. Is this reasonable or should I look at reworking this ticket to have separate org an facility slugs?

**Pull request merged**